cmd/juicify/app/juicify.go

```diff
@@ -5,66 +5,151 @@ package app
+	onConnectionError = flag.String("on-connection-error", "fail", "When a connection error happens, [fail, continue]")
+
+	juicePath = flag.String("juice-path", "", "Path to the juice executables if different than current executable path")
+
+	errCancelled            = errors.New("cancelled")
+	errQueueTimeout         = errors.New("queued GPU request timed out")
+	errInvalidConfiguration = errors.New("invalid configuration")
+)
+
+func validateConfiguration(config *Configuration) error {
+	if config.Id != "" {
+		return errInvalidConfiguration.Wrap(errors.New("id must not be specified"))
+	} else if len(config.Servers) != 0 {
+		return errInvalidConfiguration.Wrap(errors.New("servers must not be specified"))
+	} else if config.Controller == "" {
+		return errInvalidConfiguration.Wrap(errors.New("controller must be specified"))
+	} else if len(config.Requirements.Gpus) == 0 {
+		config.Requirements.Gpus = append(config.Requirements.Gpus, restapi.GpuRequirements{})
+	}
+
+	return nil
+}
+
+func requestSession(group task.Group, client *http.Client, config *Configuration) error {
+	api := restapi.Client{
+		Client:      client,
+		Address:     config.Controller,
+		AccessToken: config.AccessToken,
+	}
+
+	id, err := api.RequestSessionWithContext(group.Ctx(), config.Requirements)
+	if err != nil {
+		return err
+	}
+
+	session, err := api.GetSessionWithContext(group.Ctx(), id)
+	if err != nil {
+		return err
+	}
 
-	LogGroup string `json:"logGroup,omitempty"`
+	if session.State == restapi.SessionQueued {
+		logger.Info("Session queued")
 
-	LogFile string `json:"logFile,omitempty"`
+		timeoutChannel := make(chan struct{})
+		defer close(timeoutChannel)
 
-	ForceSoftwareDecode bool `json:"forceSoftwareDecode,omitempty"`
+		if *queueTimeout > 0 {
+			group.GoFn("Queue Timeout", func(g task.Group) error {
+				timeout := time.NewTicker(time.Duration(*queueTimeout) * time.Second)
+				defer timeout.Stop()
 
-	DisableCache bool `json:"disableCache,omitempty"`
+				select {
+				case <-group.Ctx().Done():
+					return nil
 
-	DisableCompression bool `json:"disableCompression,omitempty"`
+				case <-timeout.C:
+					timeoutChannel <- struct{}{}
+					return nil
+				}
+			})
+		}
 
-	Headless bool `json:"headless,omitempty"`
+		ticker := time.NewTicker(5 * time.Second)
+		defer ticker.Stop()
 
-	AllowTearing bool `json:"allowTearing,omitempty"`
+		for session.State == restapi.SessionQueued {
+			select {
+			case <-group.Ctx().Done():
+				err = errCancelled
 
-	FrameRateLimit int `json:"frameRateLimit,omitempty"`
+			case <-timeoutChannel:
+				err = errQueueTimeout
 
-	FramesQueueAhead int `json:"framesQueueAhead,omitempty"`
+			case <-ticker.C:
+				session, err = api.GetSessionWithContext(group.Ctx(), id)
+			}
 
-	SwapChainBuffers int `json:"swapChainBuffers,omitempty"`
+			if err != nil {
+				logger.Info("Cancelling session")
 
-	WaitForDebugger bool `json:"waitForDebugger,omitempty"`
+				api.CancelSession(id)
```

Comment:
can we also cancel the session on exit?

Reply:
The only concern with that is without the Windows Jobs, in applications like 3DMark, juicify will start a launcher which launches something else and then exits which then triggers juicify to exit. We fix that and yes we can cancel the session on juicifys exit.